Generic UI navigation system (#41)
bevyengine/rfcs: rfcs/41-ui-navigation.md

# Feature Name: Generic UI Navigation API

## Summary

Introduce [`bevy-ui-navigation`] into the bevy tree.

By default this, amounts to replacing `Interaction` with `Focusable` in `bevy_ui`.
On top of the current behavior, this adds the following capabilities:
- Focus management for **gamepads**
- Decouples input from focus management
- Decouples focus management from `bevy_ui`
- User-specified input handling for UI interaction
- An optional menu tree management system

## Terminology

- _menu tree_: the access hierarchy of menus.
- _activating_ a focusable: Sending the `Action` event while this focusable is focused.
- _focused menu_: The menu in which the currently focused `Focusable` is.
- _child menu_ (of a focusable): The menu accessed from this focusable.

> ## Motivation

Reviewer: The other big missing piece of motivation here is screenreader integration and accessibility.

`bevy_ui` interaction story is crude.
The only way to interact with the UI currently
is through the `Interaction` and `FocusPolicy` components.
`Interaction` only supports mouse or touch-based input.
We want gamepad supports and more game-oriented functionalities,
such as menu trees and movement between menus.

The current UI focus system is limited to `bevy_ui` and
can't be swapped out for something the user deems more appropriate.

> This is in contradiction with the general philosophy of bevy,

Reviewer: IMO we can just cut this paragraph.

which is of providing to the user all the tools necessary to create something
that fits their needs optimally.

> This RFC proposes to resolve all those questions.

Reviewer: Filler paragraph; cut.

[`bevy-ui-navigation`] provides a highly customizable ECS-based navigation engine.

> ## User-facing explanation

Reviewer: I think this will probably want a rewrite, but for now the bullet points are fine.

Author: I've now copied most of the API docs from bevy-ui-navigation into this RFC. I'm not a fan (just click the "see doc" link :P) but it has the quality of explaining very well each item of the library.

The [bevy-ui-navigation README][`bevy-ui-navigation`] is a good start.
To summarize, here are the headlines:
- The navigation system can only be interacted with with events:
  `NavRequest` as input and `NavEvent` as output.
- `NavEvent::FocusChanged` contains the list of entities that are traversed
  to go from the last focus position to the new one.
- ui-navigation provides default input systems, 
  but it's possible to disable them and instead use custom ones.
  This is why we restrict interactions with navigation to events.
- ui-navigation provides a system to declare buttons as "leading to menu X"
  in order to create menu trees.
- ui-navigation defines [custom `SystemParam`][event_helpers]s
  to ease reaction to events (such a button press or menu change).
  The most naive usage results into [elm architecture][elm_architecture]-like code.

>   Systems updating the model (ECS hierarchy) and systems reacting to events. 

Reviewer: Grammar is a bit weird here.

- It is possible to create isolated menus by using the `NavMenu` component.
- All `Focusable` children in the `NavMenu` entity's tree will be part
  of this menu, and moving using a gamepad from one focusable in a menu
  can only lead to another focusable within the same menu.
  You can specify a focusable in any other menu that will lead to this
  one once activated.

> ## Implementation

Reviewer: Like in the user-facing explanation, this would benefit from more prose laying out how all the bits are connected.

Author: I've added a more descriptive "Internal representation" section.

> ## Implementation

Reviewer: The other edge case I want examined is "multiple cursors", like in Smash Bros character select. This was raised in Discord and it seemed like you had a good fix, but it's not in this RFC.

Author: This might be better reserved for a "future possibilities" section. This RFC already skips a few non-necessary/nice-to-have features from `bevy-ui-navigation`, and it's still massive.

Reviewer: Agreed, I'm on board with that.

Since `bevy-ui-navigation` already exists,
we will discuss its design, it's functionalities and their uses.

Implementation details are already covered by the implementation,

> which btw is fully documented (including internal architecture).

Reviewer: ```suggestion
which is fully documented (including internal architecture).
```

### Exposed API

The crux of `bevy-ui-navigation` are the following types:

> - `Focusable` component

Reviewer: Can we model this as an enum?

Author: Interesting ideas. The current design has `Focusable` be two different enums (one enum for the state `FocusState`, and one enum for the "kind" of focusable (lock, cancel, normal)). It might be possible to separate them in two components. I've never considered this.

However, we need some sort of strategy to prevent users from directly mutating the state, because the navigation algorithm relies on having full control over it. It's read only (and for very good reason!)

>   - `lock`: activating it will lock the navigation engine and discard all `NavRequest`.

Reviewer: ```suggestion
  - `lock`: locks the navigation engine and discards all `NavRequest`.
```

>   - `cancel`: activating it will do the equivalent of receiving a `Cancel` request.

Reviewer: This feels circular and is very unclear to me.

Author: I've replaced this whole section with links to their corresponding rust doc items, since it's pointless to try to summarize the whole library into a few lines.

>   - `dormant`: this will be the preferred entity to focus when entering a menu.

Reviewer: Definitely feels like there's a missing sentence explaining what this actually does.

> - `NavMenu` enum

Reviewer: The name isn't very descriptive here; I'm not sure how this is used.

Author: Can be replace with `MenuConfig` or `MenuSettings`. I recall bevy is not very consistent when it comes to `Config`, `Configuration` or `Settings`. I actually alias it to `MenuSettings` in some places in the current implementation.

  - `Wrapping**`/`Bound**`: navigation can wrap
    if directional sticks are pressed in a direction there are no focusables in
  - `**Scope`/`**2d`: A scopped menu is tab menu navigable with special hotkeys.
- `NavRequest` event
  - `Action`: equivalent to left click or `A` button on controller
  - `Cancel`: backspace or `B` on controller
  - `FocusOn`: request to move the focus to a specific `Entity`
  - `Move(Direction)`: directional gamepad or keyboard input
  - `ScopeMove(ScopeDirection)`: tab menu hotkey
- `NavEvent` event
  - `InitiallyFocused`: We went from 0 focused to 1 focused element
  - `NoChanges`: The `NavRequest` did not change the focus state.
  - `FocusChanged`: The `NavRequest` changed the focus state,
    has the list of items that were made inactive
    and items made active.
  - `Locked`: The `NavRequest` caused the focus system to lock.
  - `Unlocked`: Something unlocked the focus system.

> - `MoveParam` trait

Reviewer: This is a odd name. Maybe something like `MenuNavigationStrategy`? It's a niche trait so we can go with something verbose and descriptive.

Author: I love this suggestion. This is a very good name

  - Used to pick a focusable in a provided direction
  - The `NavigationPlugin` is generic over the `MoveParam`
  - `DefaultNavigationPlugins` provides a default implementation of `MoveParam`

### `Focusable`

A `Focusable` is an entity that can be focused.
For gamepad-based navigation,
this requires a way to pick a focusable in a provided direction.
This is delegated to the UI library implementation
through the `MoveParam` trait.
In the case of bevy, `bevy_ui` is the UI library implementation,
and the `GlobalTransform` is used to locate focusables in regard to each-other.

The `Focusable` component holds state about what happens when it is activated
and what focus state it is in. (focused, active, inactive, etc.)

> #### Note on hovering state

Reviewer: Probably belongs in Drawbacks?

separate from the navigation library.

### Input customization

`ui-navigation` can be added to the app in two ways:
1. With `NavigationPlugin`
2. With `DefaultNavigationPlugins`

(1) only inserts the navigation systems and resources,
while (2) also adds the input systems generating the `NavRequest`s
for gamepads, mouse and keyboard.

This enables convenient defaults
while letting users insert their own custom input logic.

### What to focus first?

Gamepad input assumes an initially focused element
for navigating "move in direction" or "press action".
At the very beginning of execution, there are no focused element,
so we need fallbacks.
By default, it is any `Focusable` when none are focused yet.
But the user can spawn a `Focusable` as `dormant`
and the algorithm will prefer it when no focused nodes exist yet.

```rust
self.focusables
    .iter()
    // The focused focusable.
    .find_map(|(e, focus)| (focus.state == Focused).then(|| e))
    // Dormant focusable within the root menu (if it exists)
    .or_else(root_dormant)
    // Any dormant focusable
    .or_else(any_dormant)
    // Any focusable
    .or_else(fallback)
```

### Menu navigation handling

> The public API has `NavMenu`, but we use internally `TreeMenu`,

Reviewer: This choice isn't justified well enough. There's a fair bit of complexity here; why not just expose `NavMenu`?

Author: There are now 7 paragraphs of justification in the "Spawning menus" section.

this prevents end users from breaking assumptions about the menu trees.

> We define the `TreeMenu` component.

Reviewer: This should try to make clear that it ties into the existing `Parent` and `Child` components.

All `Focusable` direct or indirect children of a `TreeMenu` are part of that menu.
A `TreeMenu` has a `Focusable` focus parent,
when that `Focusable` is focused and the `Action` request is received,
the focus changes to the `TreeMenu`

In short, the navigation tree is built on two elements:
- The `reachable_from` that a `NavMenu` was spawned with
- The `Focusable` within the hierarchy of a `NavMenu`

> This is a tree in the ECS.

Reviewer: Cut.

A `Focusable` used to go from the root menu
to the menu in which the currently focused entity is is _active_.
The current focus position is represented as a trail of breadcrumb active
focusables and the currently focused focusable.

This is reflected into the `NavEvent::FocusChanged` event.
When going from a focusable to another,
the navigation system emits a `FocusChanged` event,
the event contains two lists of entities:
- The list of entities that after the navigation event are no more focused
  or active
- The list of entities that are newly active or focused

For the simplest case of navigating from one focusable to another
in the same menu, the two list have a single element,
so it is capital that it is easy to access those elements.
This is why we use a `NonEmpty`, from the `non_empty_vec` crate.
A `NonEmpty<T>` has a `first` method that returns a `T`,
which in our case would be the last and new focused elements.
I chose this crate over similar ones,
because the code is dead simple and easy to vet.

#### Going back to a previous menus

`Focusable`s have a `dormant` state that is set when they go from
active/focused to not active.

> This is a form of memory that allows focus to go back to the last focused

Reviewer: This is a good feature, but I'm not sure this is the clearest implementation? I think most of my problem here is that the "dormant" name isn't intuitive to me.

Author: I spent a lot of time trying to come up with a good name here.

The difficulty I had was to **not** make it sound like it is the opposite of active.
[Here is the whole enum list for reference](https://docs.rs/bevy-ui-navigation/latest/bevy_ui_navigation/enum.FocusState.html)

Reviewer: `PreviouslyActive`?

Author: I went with `Prioritized`. It doesn't tie much with the fact that it was previously active, but in practice, the variant is only used to chose which focusable to focus first in a menu, so the name makes sense.

element within a menu when it is re-visited.
This is also the mechanism used to let the user decide which focusable to focus
when none are focused yet.

#### Spawning menus

Menus store a reference to their parent,
but that parent is not necessarily their hierarchical parent.
The parent is just a button in another menu.
It is inconvenient to have to pre-spawn each button to acquire their `Entity` id
just to be able to spawn the menu you'll get to from that button.
`ui-navigation` uses a proxy component holding both the parent and the menu state info.
That proxy component is `MenuSeed`,
you can initialize it either with the `Name` or `Entity` of the parent focusable of the menu.
A system will take all `MenuSeed` components and replace them with a `TreeMenu`.

This introduces a single frame lag on menu spawn,
but it improves ergonomics to users.
I thought it was a pretty good trade-off,
since spawning menus is not time critical, unlike input.

#### cancel events

When entering a menu, you want a way to leave it and return to the previous menu.

> To do this, we have two options:

Reviewer: Why do we need both options? What's the difference?

Author: A "go back to previous menu" button is a very common pattern, and accommodating it is fairly neat.

It's not too hard for users to implement their own back button, but then it requires them to actual understand the low level navigation API we are exposing here.

Reviewer: Seems reasonable :) Explain that context here!

Our design differs from Microsoft's UI navigation in a few key ways.
We have game-specific assumptions. For example: we assume a "trail" of active elements
that leads to the sub-sub-submenu we are currently focused in, we rely on that
trail to navigate containing menus from the focused element in the submenu.

Otherwise, given my little experience in game programming, I am probably overlooking some
standard practices. My go-to source ("Game Engine Architecture" by Jason
Gregory) doesn't mention ui navigation systems.

## Open questions

### `NavRequest` system

I wanted to defer the raw input processing to the user. The current
implementation provides default systems to manage ui input, with a
`InputMapping` resource to customize a little bit the system's behaviors. 

### Game-oriented assumptions

This was not intended during implementation, but in retrospect, the `NavMenu`
system is relatively opinionated.

Indeed, we assume we are trying to build a game UI with a series of "menus" and
that it is possible to keep track of a "trail" of buttons from a "root menu" to
the deepest submenu we are currently browsing.

> The example of a fancy editor with a bunch of dockered widgets seems to break

Reviewer: ```suggestion
The example of a fancy editor with a bunch of docked widgets seems to break
```

### (Solved, but worth asking) Moving UI camera & 3d UI

The completely decoupled implementation of the navigation system
enables user to implement their own UI.
It is perfectly possible to add a `Focusable` component to a 3d object
and, for example, provide a [mouse picking] based input system
sending the `NavRequest`,
while using `world_to_viewport` in `MoveParam` for gamepad navigation.
(or completely omitting gamepad support by providing a stub `MoveParam`)

## Drawbacks and design limitations

### How does this work with the spawn/despawn workflow?

The current design imposes to the user that all UI nodes such a buttons and
menus must be already loaded, even if not visible on screen.
This is to support re-focusing the last focused element when moving between
menus.
The design of this RFC _hinges on_ the whole tree been loaded in ECS while
executing navigation requests, and would work poorly if menus were spawned and
despawned dynamically.
From what I gather from questions asked in the `#help` discord channel,
I think the design most people come up with naturally is to spawn and despawn
dynamically the menus as they are traversed, which is incompatible with
this design.

> However, in practice, in my own games, I worked around it by expanding the

Reviewer: IMO we should use RenderLayers for this instead, rather than fussing with moving the UI camera.

Author: I rather think `display: Display::None` would be the correct solution (or the simplest to implement for end users). Due to layouting and how it (does not) interact with `Visibility` and `RenderLayers`.

whole UI tree beyond the UI camera view and moving around the camera.
An alternative would simply to set the `style.display` of non-active menus to
`None` and change the style when focus enters them.
This also fixes the already-existing 1 frame latency issue with the
spawn/despawn design.

Something that automates that could be a natural extension of the exposed API.
It may also help end users go with the best design first.

> ### User upheld invariants

Reviewer: Definitely nervous about this.

Author: Honestly, not that bad. The two conditions that are absolutely necessary are checked and cause panics with a detailed error message in case the invariants are not held. In any case, if there were not enforced, it would not cause UB, just buggy behavior.

But maybe I can take a closer look at them and see if they are that necessary.

Reviewer 2: IMO there are much move severe invariants already in the engine, which to a some degree is needing to be upheld by the user (GlobalTransform + Transform for hierarchy to work comes to mind). 

That in itself is not a good argument for adding more, but we are going to need a good solution for Archetype Invariants engine-wide anyway which this design would benefit from so I'm less nervous about this. 

That being said, there is plenty of value is exploring alternatives but I don't think this should be that hard of a blocker in the name of getting some velocity on core UI features.

  to 1 relationship, see following examples)
* The navigation algorithm **resolves** a `NavRequest` by finding which `Entity`
  to next focus to
* The **active trail** is all the `Active` elements from the outermost
  `NavFence` up to the innermost `NavFence` containing the `Focused` element.

### Game developer interactions with the system

We add a _navigation tree_ to the UI system, the tree will be used to enable
key-based (and joystick controllable) navigation of UI, by allowing users to
rotate between different UI elements.

Currently, the developer must themselves write the relationship between
various buttons in the menus and specify which button press leads to other menu
items, etc. Or more simply, limit themselves to pointer-based navigation to
avoid having to think about menu navigation.

By including navigation-related components such as `Focusable` and `NavFence`,
the developer can quickly get a controller-navigable menu up and running.

We could also add the `Focusable` component to standard UI widget bundles such
as `ButtonBundle` and a potential `TextField`. This way, the developer _doesn't

> even need to specify the `Focusable` elements_.

Author: @TheRawMeatball 
> This is super exciting, though I have concerns about having people manually build the navigation tree. I think all navigation info should be derived from the existing widget tree, at no extra cost to the developer.

It is completely possible to make the widgets navigable without extra cost to the developers, simply by adding the `Focusable` component as a field to the various relevant component bundles. I didn't think of it earlier because I was locked in the "this is a plugin" mindset.

# Feature Name: Generic UI Navigation API

## Summary

A set of components and a new system to define navigation nodes for the ECS UI.

## Motivation

On top of making it easy to implement controller navigation of
in-game UIs, it enables keyboard navigations and answers accessibility concerns
such as the ones
[raised in this issue](https://github.com/bevyengine/bevy/issues/254).

## User-facing explanation

We add a _navigation tree_ to the UI system, the tree will be used to enable

> key-based navigation of the specified UIs.

Reviewer: ```suggestion
key-based (and joystick controllable) navigation of UI, by allowing users to rotate between different UI elements.
```

Currently, the developer must themselves write the relationship between
various buttons in the menus and specify which button press leads to other menu
items, etc. Or more simply, limit themselves to pointer-based navigation to
avoid having to think about menu navigation.

By including navigation-related components such as `Navigable` and `Container`,
the developer can quickly get a controller-navigable menu up and running. The

> `Container` accepts various `Plane`s to fine-tune more precisely which keys

Reviewer: Users have no idea what either a container or a plane is here, unless they've already seen them in other UI frameworks.

trigger which kind of navigations.

> To interact in real-time with the UI, the developer emits `NavCommand`s events.

Reviewer: We should avoid reusing `Command` in a similar but distinct context :)

The navigation system responds with `NavEvent` events, such as
`NavEvent::FocusChanged` or `NavEvent::Caught`.

The navigation system doesn't handle itself the state of the UI, but instead

> emits the `FocusChanged` etc. events that the developer will need to read to,

Reviewer: IMO, this shouldn't be an event. Instead, toggle the value of `Focused` enum (like we have for buttons), and then let the developers use change detection for this.

Reviewer 3: While for the current system this'd be the best option, in the context of a possibly larger reworking of UI adding a separate concept of "UI events" with propagation and using said events for this would probably be better. Out of this RFC s scope though.

Author: There is two reasons behind this:
1. It's a way for me to completely hand-wave away everything but exactly the navigation algorithm
2. It responds to that [specific request from ndarilek](https://github.com/bevyengine/bevy/issues/254#issuecomment-679359772)

  for event in events.iter() {
    if matches!(event, NavEvent::Caught(NavCommand::Action, game.ui.start_button)) {
      // start game
    }
    //etc.
  }
}
```

## Implementation strategy

### Basic physical navigation

Let's drop the idea to change submenu with LT/RT for a while. We would change
submenus by navigating upward with the Dpad or arrow keys to the tab bar and
selecting the tab with the LEFT and RIGHT keys.

Typical game engine implementation of menu navigation requires the developer to
specify the "next", "left", "right", "previous" etc.[^1] relationships between
focusable elements. This is just lame! Our game engine not only already knows
the position of our UI elements, but it also has knowledge of the logical
organization of elements through the `Parent`/`Children` relationship. (in this
case, children are specified in the parenthesis following the element name,
look at `grid(...)`)

Therefore, with the little information we provided through this hand wavy
specification, we should already have everything setup for the navigation to
just work™.

### Dimensionality

Let's go back to our ambition of using LT/RT for navigation now. 

> UI navigation is not just 2D, it's N-dimensional. The

Reviewer: Yes, firmly agree.

LT/RT to change tabs is like navigating in a 3rd orthogonal dimension to the
dimensions you navigate with UP DOWN LEFT RIGHT inside the menus.

> I'll call those dimensions `Plane`s because it's easier to type than

Reviewer: Oh I see. `Plane` is not a great word for this: it implies multidimensionality, and doesn't work for looping behavior, which is both common and important.

`Axis` is probably the most common word that could be used here.

worry about that. Our `Plane`s are:
* `Plane::Menu`: Use LT/RT to move from one element to the next/previous
* `Plane::Select`: Instead of emitting an `Action` event when left-clicking or
  pressing A/B, go up-down that direction
* `Plane::Physical`: Use the `Transform` positioning and navigate with Dpad or
  arrow keys

Each `NavCommand` moves the focus on a specific `Plane` as follow:
* `Plane::Menu`: `Previous`, `Next`
* `Plane::Select`: `Action`, `Cancel`
* `Plane::Physical`: `MoveUp`, `MoveDown`, `MoveLeft`, `MoveRight`

We should also be able to "loop" the menu, for example going LEFT of "soul"
loops us back to "all" at the other side of the tab bar.

### Specifying navigation dimensions

We posited we could easily infer the physical layout and bake a navigation map
automatically based on the `Transform` positions of the elements. This is not
the case for the dimensions of our navigation. We can't magically infer the
intent of the developer: They need to specify the dimensionality of their menus.

Let's go back to the menu. In the example code, I refer to `:container` and
`:navigable`. I've not explained yet what those are. Let's clear things up.

> A _navigable_ is an element that can be focused. A _container_ is a node entity

Reviewer: I suspect we can just reuse the existing `Interaction` component, probably reflavoring to `Focus`. The venn diagram is a circle. so we can use the same component for both marking and storing the state.

that contains _navigables_ and 0 or 1 other _container_.

![The previous tabbed menu example, but with _containers_ and _navigables_ highlighted](https://user-images.githubusercontent.com/26321040/140542768-4fdd5f23-2c2e-43c1-9fa4-cc11fe67c619.png)

The _containers_ are represented as semi-transparent squares; the _navigables_
are circles.

In rust, it might look like this:
```rust
struct Container {
  inner: Option<Box<Container>>,
  siblings: NonEmptyVec<Navigable>,
  active: SiblingIndex,
  plane: Plane,
}
```
Note: this is the data structure for the navigation algorithm, the ECS

> componenet called `Container` will probably look like this:

Reviewer: ```suggestion
component called `Container` will probably look like this:
```

```rust
#[derive(Component)]
struct Container {

>   plane: Plane,

Reviewer: So, given that we're using separate entities for these containers already, and you could have multiple of them, I think the planes (or whatever bikeshed names) should be components of the `Plane<P>` type.

Then, you can have separate systems for each plane type (using a generic helper method on `App`, stealing tech from events) and navigate very quickly using query filters. This will also make the meaning of each container much clearer in various tools like `bevy_egui_inspector`, since we don't need to go digging into the values.

Then, since looping is defined on a per-plane basis, this should be defined in the helper method on `App`.

Author: I really like this suggestion! I'll keep in mind for the implementation.

### As plugin or integrated in the engine?

I want to first draft this proposition as a standalone plugin ([see on
github](https://github.com/nicopap/ui-navigation)). The current
design doesn't require any specific insight into the engine that cannot be
accessed by a 3rd party plugin. However as I already mentioned, this solves 
[bevy-specific concerns](https://github.com/bevyengine/bevy/issues/254), and
bevy would benefit from integrating a good ui navigation system.

### Naming

I'm not comfortable with the names I chose for the various concepts I
introduced. (such a `Container`, `Plane`, `Physical`) Do not be afraid to
suggest name changes. Naming is important and I'm already not satisfied with
the nomenclature I came up with.

## Drawbacks and design limitations

### Unique hierarchy

If we implementation this with a cached navigation tree, it's going to be
easy to assume accidentally that we have a single fully reachable navigation
tree. This is not obvious, it's easy to just add the `Container` or `Navigable`
`Component` to an entity that doesn't have itself a `Container` parent more
than once.

This needs to be solved, I'm thinking that an error message would be enough in

> this case. But it's worth considering design changes that makes this situation

Reviewer: IMO this is out of scope for this RFC. This flavor of "graph invariant" is one of the central (and most challenging) problems in relations (a massive ECS-focused effort to better represent entity-entity connections).

a non-issue.

### No concurrent submenus

You have to "build up" reactively the submenus as you navigate through them,
since **there can only be a single path through the hierarchy**.

We expect that on `FocusChanged`, the developer adds the components needed if
the focus change leads to a new sub-menu.

I'm implementing first to see how cumbersome this is, and maybe revise the
design with learning from that.

Maybe it's possible to integrate more than one `Container` within other
containers. The only motivation for this currently is that it makes it easier to
implement and reason about the system.

### Inflexible navigation

> The developer might want to add arbitrary "jump" relations between various

Reviewer: IMO this should be handled by overriding the existing logic in some form, rather than something we build in an escape hatch for.

Building something in makes it much harder to teach, write and maintain and doesn't actually increase flexibility. The ECS is already an extremely powerful tool for this sort of customization.

menus and buttons. This is not possible with this design. Maybe we could add a
```rust
  bridges: HashMap<NavCommand, Entity>,
```
field to the `Container` to reroute `NavCommand`s that failed to change focus
within the menu. I think this might enable cycling references within the
hierarchy and break useful assumptions.

## Unresolved questions

Building the physical navigation tree seems not-so trivial, I'm not sure how
feasible it is, although I really think it's not that difficult.

> ## Isn't this overkill?

Reviewer: Absolutely not. I think this is about as simple as we can get while still allowing something properly robust to complex user needs.

And using two seperate frameworks (one for simple cases, and then an incompatible one for complex ones) is a terrible idea due to the fragmentation and migration pain it induces.

The tab "soul" and the panel "abc" are highlighted to show the player what menu
they are down, they are **active** elements. The circle "B" is highlighted
differently to show it is the **focused** element.

You can move `UP`/`DOWN` in the submenu, but you can also use `LT`/`RT` to
go from one submenu to the other (the _soul_, _body_ etc tabs). In code, we
might react to focus changes (for example to change the color of
focused/unfocused elements) as follow:
```rust
fn setup_ui() {
  // setup UI like you would setup UI in bevy today
}
fn handle_nav_events(mut events: EventReader<NavEvent>, game: Res<Gameui>) {
    use bevy_ui_navigation::{NavEvent::Caught, NavRequest::Action};
    for event in events.iter() {
        match event {
            Caught { from, request: Action } if from.first() == game.start_game_button => {
                  // Start the game on "A" or "ENTER" button press
            }
            _ => {}
        }
    }
}
fn button_system(
    materials: Res<Materials>,
    mut focusables: Query<(&Focusable, &mut Handle<ColorMaterial>), Changed<Focusable>>,
) {
    for (focus_state, mut material) in focusables.iter_mut() {

>         if focus_state.is_focused() {

Reviewer: Does this make more sense as a `match` on an enum? I tend to find that to be clearer, and it helps me catch edge cases more easily.

Author: It's implemented as an `enum` under the hood! Makes total sense to expose it then.

  `NavRequest::MenuMove` even when focusing in another menu.
* `NavRequest::MenuMove`: bad naming, given everything is moving within a menu
* `NavEvent::Caught`: doesn't really express that it's just a `NavRequest` that
  led to no change in focus

### `NavRequest` system

I wanted to defer the raw input processing to the user. The current
implementation **requires** the user to specify the input button mapping to the
`NavRequest`s. However, if we want to provide an out-of-the-box working
experience, we should have at least a default mapping that can be modified at
runtime.

I think it makes sense to have a default mapping, as technically it's already
the case with the `Interaction` component. And having something working
out of the box is a huge positive.

I am against an inflexible system where it is impossible to change the input
mappings, any respectable game let you change the input mapping. And if not,
your players start very quickly to ask for one! So it is necessary to be able
to remap the buttons.

How would that work? Is there already examples in bevy where we provide default
button mappings that can be changed?

> I think we should think deeply about input handling in default plugins.

Reviewer: Agreed; this should be subsumed under whatever input mapping standard we agree upon.

reachable from where. Not the relation between each individual focusable. This
drastically simplifies the structure of the tree. Because a game may have many
interactive UI elements, but not that many menus.

The previous graph diagram may have been a bit misleading. It's not the real
_navigation graph_ (that we introduced as _menu graph_ for good reasons).
The tree that the algorithm deals with and traverse is between menus
rather than `Focusable` elements. The graph as it is used in our code is the
following:

![RPG menu demo split between menus with pointers leading to the "parent
focusable"](https://user-images.githubusercontent.com/26321040/141672009-e023855e-10a7-4c50-a8dd-eff10acb2208.png)

We may also imagine an implementation where the non-selected menus are hidden
but loaded into the ECS. In this case, the navigation tree might look like
this:

![RPG menu graph without focusables, only
menus](https://user-images.githubusercontent.com/26321040/141672033-8bd43660-78af-4521-b367-03c7270c58c5.png)

### Algorithm

The algorithm is the `resolve` function in [ui-navigation](https://github.com/nicopap/ui-navigation/blob/master/src/lib.rs#L414).

> The meat of it is 50 lines of code.

Author: This is now false. The `resolve` function is 111 lines of code now

# Feature Name: Generic UI Navigation API

## Summary

Introduce [`bevy-ui-navigation`] into the bevy tree.

By default this, amounts to replacing `Interaction` with `Focusable` in `bevy_ui`.
On top of the current behavior, this adds the following capabilities:
- Focus management for **gamepads**
- Decouples input from focus management
- Decouples focus management from `bevy_ui`
- User-specified input handling for UI interaction
- An optional menu tree management system

## Terminology

- _menu tree_: the access hierarchy of menus.
- _activating_ a focusable: Sending the `Action` event while this focusable is focused.
- _focused menu_: The menu in which the currently focused `Focusable` is.
- _child menu_ (of a focusable): The menu accessed from this focusable.
- _transitive_: If `A` is _foo_ of `B` (swap _foo_ with anything such as
  _child_, _left_ etc.)
  Then `A` is _transitive foo_ of `B` if there is any chain of `C`, `D`,

>   `E` etc. such as `A` is _foo_ `C`, `C` is _foo_ `D`, ... , `E` is _foo_ `B`.

Author: I'm so sorry. I hope this is a good enough explanation for everyone :D

    pub wrapping: bool,
    /// Whether this is a scope menu.
    ///
    /// A scope menu is controlled with [`NavRequest::ScopeMove`]
    /// even when the focused element is not in this menu, but in a submenu
    /// reachable from this one.
    pub scope: bool,
}
```

#### Going back to a previous menus

`Focusable`s have a `prioritized` state that is set when they go from
active/focused to not active.
This is a form of memory that allows focus to go back to the last focused
element within a menu when it is re-visited.
This is also the mechanism used to let the user decide which focusable to focus
when none are focused yet.

#### Spawning menus

Menus store a reference to their parent,
but that parent is not necessarily their hierarchical parent.
The parent is just a button in another menu.

It is inconvenient to have to pre-spawn each button to acquire their `Entity` id
just to be able to spawn the menu you'll get to from that button.
[`bevy-ui-navigation`] uses a proxy holding both the parent
and the menu state info.

That proxy component is `MenuSeed`,
you can initialize it either with the `Name` or `Entity` of the parent focusable of the menu.

> A system will take all `MenuSeed` components and replace them with a `TreeMenu`.

Author: Note that the current implementation has simplified this. `MenuSetting` is now a component, and we have another `MenuBuilder` component that still gets converted by the navigation plugin into a `TreeMenu`.

Issue is I don't like "`MenuBuilder`" and wait to have a better name before updating the RFC.

panic fairly quickly if you don't follow the invariants.
* Invariant (1) panics as soon as you add the menu without focusable
  children.
* Invariant (2) panics if the focus goes into a menu loop.

#### Variants

* **Bound2d**: Non-wrapping menu with 2d navigation.
  It is possible to move around this menu in all cardinal directions, the
  focus changes according to the physical position of the
  [`Focusable`] in it.
  \
  If the player moves to a direction where there aren't any focusables,
  nothing will happen.
* **Wrapping2d**: Wrapping menu with 2d navigation.
  It is possible to move around this menu in all cardinal directions, the
  focus changes according to the physical position of the
  [`Focusable`] in it.
  \
  If the player moves to a direction where there aren't any focusables,
  the focus will "wrap" to the other direction of the screen.
* **BoundScope**: Non-wrapping scope menu
  Controlled with `NavRequest::ScopeMove`
  even when the focused element is not in this menu, but in a submenu
  reachable from this one.
* **WrappingScope**: Wrapping scope menu
  Controlled with `NavRequest::ScopeMove` even

>   when the focused element is not in this menu, but in a submenu reachable from this one.

Author: I'm aware this could be replace by two booleans. And I should probably do it.

("B") is `FocusState::Focused`.

Since the "tabs menu" doesn't have an outgoing arrow, it is the root menu.

The active buttons are a breadcrumb of buttons to activate to reach
the current menu with the `Focused` focusable from the root menu.

To move the focus from the "soul menu" to the "ABC menu",
you need to send `NavRequest::Action` while the button "abc" is focused
(i.e.: _"activating"_ the button).

Such a move would generate the following `NavEvent`:
```rust
NavEvent::FocusChanged {
  to: [<B entity>, <abc entity>],
  from: [<abc entity>],
}
```

If there were a "KFC menu" (currently hidden) [child of](#Terminology)
the "kfc" button, then [activating](#Terminology)
the "kfc" button would send the focus to
the prioritized focusable within the "KFC menu".

To navigate from "ABC menu" back to "soul menu",
you would send `NavRequest::Cancel`.
Such a move would generate the following `NavEvent`:
```rust
NavEvent::FocusChanged {
  to: [<abc entity>],
  from: [<B entity>, <abc entity>],
}
```
The "tabs menu" is defined as a "scope menu", which means that
by default, the `LT` and `RT` gamepad buttons will navigate

> the "tabs menu" regardless of the current focus position.

Reviewer 2: I'm guessing this is just for illustrative purposes, but we should aim to separate button inputs from action events, which would be out of scope for this RFC (and I suspect @alice-i-cecile's input manager is making a good job of that) but could this be example be rewritten to use something like `NavRequest::Next` and `NavRequest::Previous` (as suggested above)?

1. A menu must have **at least one** [`Focusable`] child in the UI
   hierarchy.
2. There must not be a menu loop. I.e.: a way to go from menu A to menu B and
   then from menu B to menu A while never going back.

#### Panics

Thankfully, programming errors are caught early and you'll probably get a
panic fairly quickly if you don't follow the invariants.
* Invariant (1) panics as soon as you add the menu without focusable
  children.
* Invariant (2) panics if the focus goes into a menu loop.

#### Fields

```rust
pub struct MenuSetting {
    /// Whether to wrap navigation.
    ///
    /// When the player moves to a direction where there aren't any focusables,
    /// if this is true, the focus will "wrap" to the other direction of the screen.
    pub wrapping: bool,
    /// Whether this is a scope menu.
    ///
    /// A scope menu is controlled with [`NavRequest::ScopeMove`]
    /// even when the focused element is not in this menu, but in a submenu
    /// reachable from this one.
    pub scope: bool,
}
```

#### Going back to a previous menus

`Focusable`s have a `prioritized` state that is set when they go from
active/focused to not active.

> This is a form of memory that allows focus to go back to the last focused

Reviewer 2: I think we need the ability for the developer to define: 

"When coming back to this menu, should we focus on the last-focused element or should we focus on the initial default element"

There are use cases in apps (filling in a new form) or games where we want to treat a menu as "first-time-user-visits" even though the user has visited before. The FFX battle system also springs to mind where you can choose between these two modes, which has a profound effect on muscle-memorisation.

Author: That's something I overlooked. Current design doesn't allow the user to modify the `prioritized` focusable. I think this can be enabled by providing a custom `SystemParam`.

It is simply added back as a component independent
from the focus system.

Note that the new `Hover` component does not comply with `FocusPolicy`.
This allows reducing the complexity of the focus code.

#### `Clicked` is not a focus state

Compared to the current `Interaction` system, the new focus system
separates semantics of "action" events and "UI focus state".

Meaning that `Interaction::Clicked`'s equivalent in the new navigation system
is the `NavEvent::NoChanges` event.

This is important, as code relying on the "Clicked" focus state
will need to be replace with reaction to a `NavEvent`.

Looking at the code changes in the existing UI examples in [this RFC's PR]
is a good indicator at how the code will need to be changed.

### How does this work with the spawn/despawn workflow?

The current design imposes to the user that all UI nodes such a buttons and
menus must be already loaded, even if not visible on screen.
This is to support re-focusing the last focused element when moving between
menus.
The design of this RFC _hinges on_ the whole tree been loaded in ECS while
executing navigation requests, and would work poorly if menus were spawned and

> despawned dynamically.

Reviewer 2: My mind keeps coming back to this statement. It is an extremely common use-case to not know the data ahead of time: I want to select a villager from a list. The game queries the list of villagers currently available, and spawns buttons for them so that I can click on one. Or just a search bar in an application.

Does this mean that we cannot support those use cases at all?

Reviewer: Agreed, now that you highlight this I share your concerns.

Author: So. To  be precise, this section is really meant to contrast with the pattern of spawning and despawning each menu as they are created.

The limitation here (I've not tested it) I think is only when despawning the parent menu (for example despawning the main menu when entering the options menu), not when creating dynamically new menus you can enter, or despawning menus that do not have focus.

When you go back one level of menus (pressing backspace or clicking a "back" button), the navigation algo reads the `Entity` of the button the menu is "reachable from" (aka the button that activated the menu), and if that `Entity` is despawned, probably bad things happen (panic! most likely), as we now sort of "dereference a dangling `Entity`".

However, the reverse case of creating new menus at runtime is widely supported, the `MenuBuilder::ParentEntity` enum variant should help with runtime spawning of new menus, since you can just give it whatever `Entity` with a `Focusable` component you have available. It does mean however that the submenus must be created before the button is activated (for example when it gets focused/hovered).

It is not necessary for the whole tree to be loaded at once, but rather than the path back to the root menu is. For support for giving back focus to the last focused element (prioritized) you would also need to keep the `Focusable`s and menu tree of already-explored menus loaded. But typically, for a dropdown, this isn't something people expect.

This is why I mentioned an "infinitely deep recursive menu" as an edge case ui-navigation couldn't support.

I'm currently making a sort of "infinite weapon upgrade menu" as a cool example to demonstrate how easy it is to use your own navigation methods (using `Sprite` and `Text2d`) and generating new menus on the fly.

Author: When it comes to creating new `Focusable`s in an existing menu, it should be alright. But it's always going to be difficult when removing them (beside the mandatory minimum 1 `Focusable` in a menu). `TreeMenu` has a `active_child` field for avoiding iterating all children, but it completely violates the single-source-of-truth principle which makes despawning `Focusable`s dangerous (you must not despawn a focused or active focusable)

I'll probably have to rework access to the `active_child` to fallback to the ECS hierarchy when it doesn't exist anymore.

Reviewer 2: That is good to know, and still impressive work with all of this. Is the last remaining hurdle then to find a good way to remove _leaf_ nodes in the navigation tree dynamically?

The requirement that the path back to root must exist I think is acceptable as long as we can dynamically remove leaf nodes
